File: libUtil/src/main/java/org/waynezhou/libUtil/EnumClass.java
package org.waynezhou.libUtil;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public abstract class EnumClass<T> {
    public final T identifier;
    public final String statement;
    private Class<T> clz;
    protected EnumClass(Class<T> clz, T identifier, String statement) {
        this.clz = clz;
        this.identifier = identifier;
        this.statement = statement;
    }

    @Override
    public int hashCode() {
        return identifier.hashCode();
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if(obj==null) return false;
        if(!(obj instanceof EnumClass)) return false;
        if((((EnumClass<?>) obj).identifier.getClass() != clz)) return false;

        //noinspection unchecked
        return identifier.equals(((EnumClass<T>) obj).identifier);
    }

    @NonNull
    @Override
    public String toString() {
        return statement.toString();
    }

    public static abstract class Int extends EnumClass<Integer>{

        protected Int(int identifier, String statement) {
            super(Integer.class, identifier, statement);
        }
        protected Int(String statement) {
            super(Integer.class, statement.hashCode(), statement);
        }
    }
}

File: libUtilKt/src/main/java/org/waynezhou/libutilkt/enum/EnumClass.kt
package org.waynezhou.libutilkt.enum



@Suppress("LeakingThis", "MemberVisibilityCanBePrivate")
abstract class EnumClass<T: Any, TEnumClass:EnumClass<T,TEnumClass>> protected constructor(
    private val clz: Class<T>,
    val identifier: T,
    val statement: String
) {
    companion object{
        private val maps: MutableMap<Class<EnumClass<*,*>>, MutableMap<Any, Any>> = mutableMapOf()
        @Suppress("UNCHECKED_CAST")
        fun <T:Any, TEnumClass:EnumClass<T, TEnumClass>> getEnumMap(clz:Class<EnumClass<T, TEnumClass>>): MutableMap<T, TEnumClass>? {
            return maps[clz as Class<EnumClass<*,*>>] as? MutableMap<T, TEnumClass>
        }
    }
    init {
        val clzMap = maps.getOrDefault(this.javaClass, mutableMapOf())
        clzMap[identifier] = this
        maps[javaClass] = clzMap
    }
    override fun hashCode(): kotlin.Int {
        return identifier.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (other.javaClass != this.javaClass) return false
        return identifier == (other as? EnumClass<*,*>)?.identifier
    }

    override fun toString(): String {
        return statement
    }

    abstract class Int(statement: String, identifier: kotlin.Int) :
        EnumClass<kotlin.Int, Int>(kotlin.Int::class.java, identifier, statement){
            constructor(statement: String): this(statement, statement.hashCode())
        }

}
